### ISSUE TITLE
bug when using api trade

### ISSUE THREAD

**Issue by vibros68** [decred/dcrdex#3003]
I am trying to use bison via api endpoint. Testing with trade api and get this error:

* api endpoint: http://127.0.0.1:5758/api/trade
* method: POST
* error: Post "http://127.0.0.1:5758/api/trade": EOF
* body: 
```{"host":"dex.decred.org:7232","isLimit":true,"sell":true,"base":42,"quote":966,"qty":15000000,"rate":32000000000,"tifnow":false,"options":null}```
* bison log:
```
2024-10-02 08:54:41.131 [DBG] WEB[WS]: New websocket client 127.0.0.1:48414
2024/10/02 15:54:41 http: panic serving 127.0.0.1:48436: runtime error: slice bounds out of range [-1:]
goroutine 414 [running]:
net/http.(*conn).serve.func1()
	net/http/server.go:1947 +0xbe
panic({0x1c7ac00?, 0xc0155b8030?})
	runtime/panic.go:785 +0x132
github.com/go-chi/chi/v5/middleware.prettyStack.decorateFuncCallLine({}, {0xc006b03b26, 0x1f}, 0x1, 0x8)
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:130 +0x525
github.com/go-chi/chi/v5/middleware.prettyStack.decorateLine({}, {0xc006b03b26?, 0x111c?}, 0x1, 0x8)
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:106 +0x154
github.com/go-chi/chi/v5/middleware.prettyStack.parse({}, {0xc001110000, 0x111c, 0xc001068e98?}, {0x1ac7e20, 0x3621cc0})
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:89 +0x3bc
github.com/go-chi/chi/v5/middleware.PrintPrettyStack({0x1ac7e20, 0x3621cc0})
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:46 +0x3b
github.com/go-chi/chi/v5/middleware.(*defaultLogEntry).Panic(0x43b385?, {0x1ac7e20?, 0x3621cc0?}, {0x43b3c0?, 0xc000e061c0?, 0xc001068fe0?})
	github.com/go-chi/chi/v5@v5.0.1/middleware/logger.go:165 +0x25
github.com/go-chi/chi/v5/middleware.Recoverer.func1.1()
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:28 +0xc8
panic({0x1ac7e20?, 0x3621cc0?})
	runtime/panic.go:785 +0x132
decred.org/dcrdex/client/core.(*Core).prepareTradeRequest(0xc0000fd600, {0xc0004e23e0?, 0xc006b2e200?, 0xc0009fefc0?}, 0x0)
	decred.org/dcrdex@v1.0.0/client/core/core.go:6112 +0x4a
decred.org/dcrdex/client/core.(*Core).Trade(0xc0000fd600, {0xc0004e23e0?, 0xc015448b40?, 0x198c220?}, 0xc0011ea300?)
	decred.org/dcrdex@v1.0.0/client/core/core.go:5772 +0x1d
decred.org/dcrdex/client/webserver.(*WebServer).apiTrade(0xc0003f04b0, {0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	decred.org/dcrdex@v1.0.0/client/webserver/api.go:632 +0x210
net/http.HandlerFunc.ServeHTTP(0xc0003f04b0?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0x8c7d5d?)
	net/http/server.go:2220 +0x29
decred.org/dcrdex/client/webserver.(*WebServer).rejectUnauthed-fm.(*WebServer).rejectUnauthed.func1({0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	decred.org/dcrdex@v1.0.0/client/webserver/middleware.go:132 +0x68
net/http.HandlerFunc.ServeHTTP(0xc0009afae0?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0xc01b212120?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5.(*ChainHandler).ServeHTTP(0x1aacc80?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0xc000bb14e0?)
	github.com/go-chi/chi/v5@v5.0.1/chain.go:31 +0x26
github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0009a7b60, {0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	github.com/go-chi/chi/v5@v5.0.1/mux.go:437 +0x1f2
net/http.HandlerFunc.ServeHTTP(0x1aa03c0?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0x10?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5/middleware.AllowContentType.func1.1({0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	github.com/go-chi/chi/v5@v5.0.1/middleware/content_type.go:41 +0xdd
net/http.HandlerFunc.ServeHTTP(0xc0154317f8?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0xc0154317d8?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0009a7b60, {0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	github.com/go-chi/chi/v5@v5.0.1/mux.go:71 +0x32f
github.com/go-chi/chi/v5.(*Mux).Mount.func1({0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	github.com/go-chi/chi/v5@v5.0.1/mux.go:312 +0x1bb
net/http.HandlerFunc.ServeHTTP(0x1aacc80?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0xc01563013d?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0005f2ae0, {0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	github.com/go-chi/chi/v5@v5.0.1/mux.go:437 +0x1f2
net/http.HandlerFunc.ServeHTTP(0xc01b212910?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0x10?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5/middleware.Recoverer.func1({0x7027ac32f478?, 0xc0009ff1c0?}, 0x1daf735?)
	github.com/go-chi/chi/v5@v5.0.1/middleware/recoverer.go:37 +0x6c
net/http.HandlerFunc.ServeHTTP(0xc000ea11a0?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0x1e77889?)
	net/http/server.go:2220 +0x29
decred.org/dcrdex/client/webserver.(*WebServer).securityMiddleware-fm.(*WebServer).securityMiddleware.func1({0x7027ac32f478, 0xc0009ff1c0}, 0xc015448b40)
	decred.org/dcrdex@v1.0.0/client/webserver/middleware.go:34 +0x174
net/http.HandlerFunc.ServeHTTP(0x362b960?, {0x7027ac32f478?, 0xc0009ff1c0?}, 0x7027ac57e078?)
	net/http/server.go:2220 +0x29
decred.org/dcrdex/client/webserver.New.RequestLogger.func3.1({0x284f080, 0xc0155b0620}, 0xc015448a00)
	github.com/go-chi/chi/v5@v5.0.1/middleware/logger.go:57 +0x154
net/http.HandlerFunc.ServeHTTP(0x28527c0?, {0x284f080?, 0xc0155b0620?}, 0x3623530?)
	net/http/server.go:2220 +0x29
github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0005f2ae0, {0x284f080, 0xc0155b0620}, 0xc0154488c0)
	github.com/go-chi/chi/v5@v5.0.1/mux.go:88 +0x2ee
net/http.serverHandler.ServeHTTP({0x284b340?}, {0x284f080?, 0xc0155b0620?}, 0x6?)
	net/http/server.go:3210 +0x8e
net/http.(*conn).serve(0xc000b60a20, {0x2852788, 0xc00065e0f0})
	net/http/server.go:2092 +0x5d0
created by net/http.(*Server).Serve in goroutine 172
	net/http/server.go:3360 +0x485
2024-10-02 08:54:41.165 [ERR] WEB[WS][127.0.0.1:48414]: Websocket receive error from peer 127.0.0.1:48414: websocket: close 1006 (abnormal closure): unexpected EOF (*websocket.CloseError)
```

**Comment by vibros68** [comment]
Oh I found out that I submitted wrong trade form body. But the server should return a message error instead of panic

**Comment by JoeGruffins** [comment]
Thanks yeah, there were a few places that could panic as we were just checking the base form and not the inner structs for nil. Attempting to fix with #3006
